File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/M._Просеивание_вверх/solution.cpp
#include <vector>
#include <cassert>
#ifdef REMOTE_JUDGE
#include "solution.h"
#endif


int siftUp(std::vector<int>& heap, int idx) {
    /*
    Time Complexity O(H) -> O(logN)
    Memory Complexity O(H) -> O(logN)
    */
    if (idx == 1) return idx;
    int parentIdx = idx / 2;
    if (heap[parentIdx] < heap[idx]) {
        std::swap(heap[parentIdx], heap[idx]);
        return siftUp(heap, parentIdx);
    }
    return idx;
}

#ifndef REMOTE_JUDGE
void test() {
    std::vector<int> sample = {-1, 12, 6, 8, 3, 15, 7};
    assert(siftUp(sample, 5) == 1);
}


int main() {
    test();
}
#endif

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/M._Просеивание_вверх/solution.py
def sift_up(heap, idx) -> int:
    """
    Time Complexity O(H) -> O(logN)
    Memory Complexity O(H) -> O(logN)
    """
    if idx == 1:
        return idx
    parent_idx: int = idx // 2
    if heap[idx] > heap[parent_idx]:
        heap[idx], heap[parent_idx] = heap[parent_idx], heap[idx]
        return sift_up(heap, parent_idx)
    return idx


def test():
    sample = [-1, 12, 6, 8, 3, 15, 7]
    assert sift_up(sample, 5) == 1


if __name__ == '__main__':
    test()